pkg/volume/csi/csi_test.go

```diff
@@ -0,0 +1,339 @@
+	utiltesting "k8s.io/client-go/util/testing"
+	"k8s.io/kubernetes/pkg/features"
+	"k8s.io/kubernetes/pkg/volume"
+	volumetest "k8s.io/kubernetes/pkg/volume/testing"
+)
+
+// TestCSI_VolumeAll runs a close approximation of volume workflow
+// based on operations from the volume manager/reconciler/operation executor
+func TestCSI_VolumeAll(t *testing.T) {
+	defer utilfeaturetesting.SetFeatureGateDuringTest(t, utilfeature.DefaultFeatureGate, features.CSIInlineVolume, true)()
+	tests := []struct {
+		name     string
+		specName string
+		driver   string
+		volName  string
+		specFunc func(specName, driver, volName string) *volume.Spec
+		podFunc  func() *api.Pod
+	}{
+		{
+			name:     "PV with reconstruction",
+			specName: "pv2",
+			driver:   "simple-driver",
+			volName:  "vol2",
+			specFunc: func(specName, driver, volName string) *volume.Spec {
+				return volume.NewSpecFromPersistentVolume(makeTestPV(specName, 20, driver, volName), false)
+			},
+			podFunc: func() *api.Pod {
+				podUID := types.UID(fmt.Sprintf("%08X", rand.Uint64()))
+				return &api.Pod{ObjectMeta: meta.ObjectMeta{UID: podUID, Namespace: testns}}
+			},
+		},
+		{
+			name:    "volume source with reconstruction",
+			driver:  "inline-driver-1",
+			volName: "test.vol2",
+			specFunc: func(specName, driver, volName string) *volume.Spec {
+				return volume.NewSpecFromVolume(makeTestVol(specName, driver))
+			},
+			podFunc: func() *api.Pod {
+				podUID := types.UID(fmt.Sprintf("%08X", rand.Uint64()))
+				return &api.Pod{ObjectMeta: meta.ObjectMeta{UID: podUID, Namespace: testns}}
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			tmpDir, err := utiltesting.MkTmpdir("csi-test")
+			if err != nil {
+				t.Fatalf("can't create temp dir: %v", err)
+			}
+			defer os.RemoveAll(tmpDir)
+			client := fakeclient.NewSimpleClientset()
+			fakeWatcher := watch.NewRaceFreeFake()
+
+			factory := informers.NewSharedInformerFactory(client, csiResyncPeriod)
+			go factory.Start(wait.NeverStop)
+
+			host := volumetest.NewFakeVolumeHostWithCSINodeName(
+				tmpDir,
+				client,
+				nil,
+				"csi-node",
+				factory.Storage().V1beta1().CSIDrivers().Lister(),
+			)
+
+			plugMgr := &volume.VolumePluginMgr{}
+			plugMgr.InitPlugins(ProbeVolumePlugins(), nil /* prober */, host)
+			csiClient := setupClient(t, true)
+
+			volSpec := test.specFunc(test.specName, test.driver, test.volName)
+			pod := test.podFunc()
+			attachName := getAttachmentName(test.volName, test.driver, string(host.GetNodeName()))
+			t.Log("csiTest.VolumeAll starting...")
+
+			// *************** Attach/Mount volume resources ****************//
+			// attach volume
+			t.Log("csiTest.VolumeAll Attaching volume...")
+			attachPlug, err := plugMgr.FindAttachablePluginBySpec(volSpec)
+			if err != nil {
+				t.Fatalf("csiTest.VolumeAll PluginManager.FindAttachablePluginBySpec failed: %v", err)
+			}
+
+			var devicePath string
+			if attachPlug != nil {
```

Comment:
Can we add a parameter to the test case whether or not we expect the volume to be attachable or device mountable?

That way we can fail the test if we ended up skipping one of them when we shouldn't have

Reply:
ok, agreed.